cfile/cfile.cpp

```diff
@@ -394,363 +363,165 @@ CFILE *open_file_in_lib(const char *filename) {
-      if (!stricmp(namebuffer, real_file)) {
-        // we found a match!
-        found_match = true;
-        break;
-      }
-    }
-    ff.Close();
-
-    if (found_match) {
-      strcpy(new_filename, namebuffer);
-      mprintf(1, "CFILE: Using \"%s\" instead of \"%s\"\n", new_filename, real_file);
-      break;
-    }
+  if (found != end(it)) {
+    // Match, append to result
+    result = found->path();
+    search_path = result;
+  } else {
+    // Component not found, mission failed
+    return {};
   }
 
-  return found_match;
+  return result.filename();
 }
 
-FILE *open_file_in_directory_case_sensitive(const char *directory, const char *filename, const char *mode,
-                                            char *new_filename) {
-  char t_dir[_MAX_PATH];
-  ddio_SplitPath(filename, t_dir, nullptr, nullptr);
-  if (cf_FindRealFileNameCaseInsenstive(directory, filename, new_filename)) {
-    // we have a file, open it open and use it
-    char full_path[_MAX_PATH * 2];
-    // if we had a directory as part of the file name, put it back in
-    if (strlen(t_dir) > 0)
-      directory = t_dir;
-    if (directory != nullptr) {
-      ddio_MakePath(full_path, directory, new_filename, NULL);
-    } else {
-      strcpy(full_path, new_filename);
-    }
-
-    return fopen(full_path, mode);
-  }
-  return nullptr;
-}
-#endif
-
 // look for the file in the specified directory
-static CFILE *open_file_in_directory(const char *filename, const char *mode, const char *directory);
+static CFILE *open_file_in_directory(const std::filesystem::path &filename, const char *mode,
+                                     const std::filesystem::path &directory);
 
 // look for the file in the specified directory
-CFILE *open_file_in_directory(const char *filename, const char *mode, const char *directory) {
+CFILE *open_file_in_directory(const std::filesystem::path &filename, const char *mode,
+                              const std::filesystem::path &directory) {
   FILE *fp;
   CFILE *cfile;
-  char path[_MAX_PATH * 2];
+  std::filesystem::path using_filename;
   char tmode[3] = "rb";
-  if (directory != nullptr) {
+  if (std::filesystem::is_directory(directory)) {
     // Make a full path
-    ddio_MakePath(path, directory, filename, NULL);
-  } else // no directory specified, so just use filename passed
-    strcpy(path, filename);
+    using_filename = directory / filename;
+  } else {
+    // no directory specified, so just use filename passed
+    using_filename = filename;
+  }
+
   // set read or write mode
   tmode[0] = mode[0];
-  // if mode is "w", then open in text or binary as requested.  If "r", alsway open in "rb"
+  // if mode is "w", then open in text or binary as requested.  If "r", always open in "rb"
   tmode[1] = (mode[0] == 'w') ? mode[1] : 'b';
   // try to open file
-  fp = fopen(path, tmode);
+  fp = fopen(using_filename.u8string().c_str(), tmode);
```

Comment:
How do you know that UTF-8 is the correct encoding to use for `fopen()`’s `filename` parameter?

Reply:
This way how std::filesystem::path represent native encoding on systems. POSIX filesystem is native `char` representation, but Windows' native representation is `wchar`. If you want use paths with C functions, you need convert them with `c_str()` using `u8string()`.

Reply:
> This way how std::filesystem::path represent native encoding on systems. POSIX filesystem is native `char` representation, but Windows' native representation is `wchar`. If you want use paths with C functions, you need convert them with `c_str()` using `u8string()`.

I’m a little bit confused. [According to cppreference.com, `std::filesystem::path::u8string()` returns a `std::u8string`](https://en.cppreference.com/w/cpp/filesystem/path/string). [`std::u8string` is the same as `std::basic_string<char8_t>`.](https://en.cppreference.com/w/cpp/string/basic_string) [`std::basic_string::<whatever>::c_str()` returns a `whatever *`](https://en.cppreference.com/w/cpp/string/basic_string/c_str) which means that `my_u8string.c_str()` should return a `char8_t *`, right? It seems to me like `using_filename.u8string().c_str()` is always going to evaluate to a `char8_t *`. I just don’t see how it could evaluate to a `wchar_t *`.

(The `std::u8string` type was added in C++20. In C++17, a similar argument can be made, but you have to replace `std::u8string` with `std::string` and `char8_t` with `char`.)

Passing a UTF-8 encoded string to `fopen()` is potentially a problem. [According to Microsoft’s documentation,](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-wfopen?view=msvc-170#remarks)

> The **`fopen`** function opens the file specified by *`filename`*. By default, a narrow *`filename`* string is interpreted using the ANSI codepage (`CP_ACP`). In Windows Desktop applications, it can be changed to the OEM codepage (`CP_OEMCP`) by using the [`SetFileApisToOEM`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-setfileapistooem) function. You can use the [`AreFileApisANSI`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-arefileapisansi) function to determine whether *`filename`* is interpreted using the ANSI or the system default OEM codepage.

Is there anything that’s going to guarantee that we’re using a UTF-8 codepage? [Windows won’t use a UTF-8 codepage by default unless the user has “Beta: Use Unicode UTF-8 for worldwide language support” enabled, or the application explicitly specifies a UTF-8 codepage.](https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page)

Additionally, if `using_filename.u8string().c_str()` sometimes does evaluate to a `wchar_t *`, then I think that we have a problem. [According to Microsoft’s documentation, you’re supposed to pass a `char *` to `fopen()`, not a `wchar_t *`. If you want to use a `wchar_t *` for the `filename` parameter, then you’re supposed to call `_wfopen()`.](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-wfopen?view=msvc-170)

Reply:
We don't even interested here what encoding is actually in filesystem. We effectively receive C-string (and don't bother in what encoding is it, it's just sequence of chars) and we ensuring that C-string is actually valid path in native filesystem. We do not using UTF-8 here at all. It's just way to convert native encoding to C-string.

Reply:
> We don't even interested here what encoding is actually in filesystem.

If we’re going to call `fopen()` on Windows, then I think that we need to be interested in the `filename` parameter’s encoding because Windows is interested in the `filename` parameter’s encoding.

---

> We effectively receive C-string (and don't bother in what encoding is it, it's just sequence of chars) and we ensuring that C-string is actually valid path in native filesystem. We do not using UTF-8 here at all.

That last sentence is very surprising to me. I thought that `std::filesystem::path::u8string` was supposed to return UTF-8, that `std::filesystem::path::u16string` was supposed to return UTF-16 and that `std::filesystem::path::u32string` was supposed to return UTF-32. [cppreference.com says the following about those functions:](https://en.cppreference.com/w/cpp/filesystem/path/string)

> Returns the internal pathname in native pathname format, converted to specific string type. Conversion, if any, is performed as follows:
>
> […]
>
> 3\) The result encoding in the case of `u8string()` is always UTF-8.

Additionally, [Microsoft’s documentation for `std::filesystem::path::u8string` says:](https://learn.microsoft.com/en-us/cpp/standard-library/path-class?view=msvc-170#u8string)

> Converts the sequence stored in `mypath` to UTF-8 and returns it stored in an object of type `u8string`.

What makes you say “We do not using UTF-8 here at all”?

---

> It's just way to convert native encoding to C-string.

I agree. `u8string().c_str()` does indeed convert a `path` (which uses its native encoding internally) into a C-string. My concern is about the encoding of the resulting C-string.